### PR TITLE
feat(object): add `project_id` resource attribute

### LINKED ISSUE #1667: Allow users to specify the project a bucket should belong to
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Each bucket should be linked to a specific project.
Currently, it is not possible to choose said project at the bucket level; thus, we need to rely on the "default" project linked to the API key.
It would be helpful to be able to either be able to set the owner project, or at least to provide an "expected_project_id", and having terraform bail out if the bucket is not linked to the right project.

### New or Affected Resource(s)

* scaleway_object_bucket

### Potential Terraform Configuration

```hcl
resource "scaleway_object_bucket" "my_bucket" {
  name = "some-example-bucket"
  region = "fr-par"
  project_id = scaleway_account_project.some_project.id
}
```

### PR DESCRIPTION
**Pull request by quantumsheep**
Closes #1667

### CODE REVIEW

File: scaleway/resource_object_bucket_test.go [hunk 1]
```diff
@@ -591,33 +591,33 @@ func TestAccScalewayObjectBucket_Cors_Delete(t *testing.T) {
...
+// 							allowed_origins = [""]
+// 							expose_headers  = ["x-amz-server-side-encryption", "ETag"]
+// 							max_age_seconds = 3000
+// 						}
+// 					}`, bucketName),
+// 				ExpectError: regexp.MustCompile("error putting S3 CORS"),
+// 			},
+// 		},
+// 	})
+// }
```
Comment (Codelax) [inline scaleway/resource_object_bucket_test.go:620]: This should be fixed in another PR

File: docs/data-sources/object_bucket.md [hunk 2]
```diff
@@ -29,6 +29,7 @@ data "scaleway_object_bucket" "selected" {
 - `name` - (Required) The bucket name.
 - `object_lock_enabled` - (Optional) Enable object lock on the bucket. Defaults to `false`. Updating this field will force creating a new bucket.
 - `region` - (Defaults to [provider](../index.md#region) `region`) The [region](../guides/regions_and_zones.md#zones) in which the Object Storage exists.
+- `project_id` - (Defaults to [provider](../index.md#project_id) `project_id`) The ID of the project the SSH key is associated with.
```
Comment (Codelax) [inline docs/data-sources/object_bucket.md:32]:
```suggestion
- `project_id` - (Defaults to [provider](../index.md#project_id) `project_id`) The ID of the project the bucket is associated with.
```

File: docs/resources/object.md [hunk 3]
```diff
@@ -38,6 +38,7 @@ The following arguments are supported:
 * `visibility` - (Optional) Visibility of the object, `public-read` or `private`
 * `metadata` - (Optional) Map of metadata used for the object, keys must be lowercase
 * `tags` - (Optional) Map of tags
+* `project_id` - (Defaults to [provider](../index.md#project_id) `project_id`) The ID of the project the SSH key is associated with.
```
Comment (Codelax) [inline docs/resources/object.md:41]:
```suggestion
* `project_id` - (Defaults to [provider](../index.md#project_id) `project_id`) The ID of the project the bucket is associated with.
```

File: docs/resources/object_bucket.md [hunk 4]
```diff
@@ -115,6 +115,7 @@ The following arguments are supported:
 * `versioning` - (Optional) A state of [versioning](https://docs.aws.amazon.com/AmazonS3/latest/dev/Versioning.html) (documented below)
 * `cors_rule` - (Optional) A rule of [Cross-Origin Resource Sharing](https://docs.aws.amazon.com/AmazonS3/latest/dev/cors.html) (documented below).
 * `force_destroy` - (Optional) Enable deletion of objects in bucket before destroying, locked objects or under legal hold are also deleted and **not** recoverable
+* `project_id` - (Defaults to [provider](../index.md#project_id) `project_id`) The ID of the project the SSH key is associated with.
```
Comment (Codelax) [inline docs/resources/object_bucket.md:118]:
```suggestion
* `project_id` - (Defaults to [provider](../index.md#project_id) `project_id`) The ID of the project the bucket is associated with.
```

File: docs/resources/object_bucket_object_lock_configuration.md [hunk 5]
```diff
@@ -55,6 +55,8 @@ The following arguments are supported:
 
         - `years` - (Optional) The number of years that you want to specify for the default retention period.
 
+- `project_id` - (Defaults to [provider](../index.md#project_id) `project_id`) The ID of the project the SSH key is associated with.
```
Comment (Codelax) [inline docs/resources/object_bucket_object_lock_configuration.md:58]:
```suggestion
- `project_id` - (Defaults to [provider](../index.md#project_id) `project_id`) The ID of the project the bucket is associated with.
```

File: docs/resources/object_bucket_website_configuration.md [hunk 6]
```diff
@@ -70,6 +70,7 @@ The following arguments are supported:
 * `bucket` - (Required, Forces new resource) The name of the bucket.
 * `index_document` - (Required) The name of the index document for the website [detailed below](#index_document).
 * `error_document` - (Optional) The name of the error document for the website [detailed below](#error_document).
+* `project_id` - (Defaults to [provider](../index.md#project_id) `project_id`) The ID of the project the SSH key is associated with.
```
Comment (Codelax) [inline docs/resources/object_bucket_website_configuration.md:73]:
```suggestion
* `project_id` - (Defaults to [provider](../index.md#project_id) `project_id`) The ID of the project the bucket is associated with.
```

File: scaleway/data_source_object_bucket_test.go [hunk 7]
```diff
@@ -38,3 +44,191 @@ func TestAccScalewayDataSourceObjectStorage_Basic(t *testing.T) {
...
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:          func() { testAccPreCheck(t) },
+		ProviderFactories: fakeSideProjectProviders(ctx, tt, project, iamAPIKey),
+		CheckDestroy:      testAccCheckScalewayObjectDestroy(tt),
+		Steps: []resource.TestStep{
+			// Create a bucket from the main provider into the side project and read it from the side provider
+			// The side provider should only be able to read the bucket from the side project
+			{
+				Config: fmt.Sprintf(`
```
Comment (Monitob) [inline scaleway/data_source_object_bucket_test.go:67]: Please add this example on the resource doc with explanation

File: scaleway/data_source_object_bucket_test.go [hunk 8]
```diff
@@ -38,3 +44,191 @@ func TestAccScalewayDataSourceObjectStorage_Basic(t *testing.T) {
...
+	projectAPI := accountV2.NewAPI(tt.Meta.scwClient)
+	project, err := projectAPI.CreateProject(&accountV2.CreateProjectRequest{
+		Name: projectName,
+	})
+	if err != nil {
+		terminate()
+		return nil, nil, nil, err
+	}
+	terminateFunctions = append(terminateFunctions, func() {
+		err := projectAPI.DeleteProject(&accountV2.DeleteProjectRequest{
```
Comment (Monitob) [inline scaleway/data_source_object_bucket_test.go:150]: better add this to the helpers. Could be usefull on the the future. Can you define custom type instead of func()

File: scaleway/data_source_object_bucket_test.go [hunk 9]
```diff
@@ -38,3 +44,191 @@ func TestAccScalewayDataSourceObjectStorage_Basic(t *testing.T) {
...
+		err := iamAPI.DeleteAPIKey(&iam.DeleteAPIKeyRequest{
+			AccessKey: iamAPIKey.AccessKey,
+		})
+		require.NoError(t, err)
+	})
+
+	return project, iamAPIKey, terminate, nil
+}
+
+func fakeSideProjectProviders(ctx context.Context, tt *TestTools, project *accountV2.Project, iamAPIKey *iam.APIKey) map[string]func() (*schema.Provider, error) {
```
Comment (Monitob) [inline scaleway/data_source_object_bucket_test.go:210]: Please add this to the provider_test helpers

File: scaleway/data_source_object_bucket_test.go [hunk 10]
```diff
@@ -38,3 +44,191 @@ func TestAccScalewayDataSourceObjectStorage_Basic(t *testing.T) {
...
+		httpClient:          tt.Meta.httpClient,
+		forceProjectID:      project.ID,
+		forceOrganizationID: project.OrganizationID,
+		forceAccessKey:      iamAPIKey.AccessKey,
+		forceSecretKey:      *iamAPIKey.SecretKey,
+	})
+	require.NoError(t, err)
+
+	providers := map[string]func() (*schema.Provider, error){
+		"side": func() (*schema.Provider, error) {
```
Comment (Monitob) [inline scaleway/data_source_object_bucket_test.go:224]: Can you add a comment for this alias.

File: scaleway/testdata/provider-instance-ip-zones.cassette.yaml [hunk 11]
```diff
@@ -230,24 +212,21 @@ interactions:
...
         terraform/terraform-tests
-    url: https://api.scaleway.com/instance/v1/zones/fr-par-2/ips/07eae38f-a230-4586-9d58-1d1476f2d07b
+    url: https://api.scaleway.com/instance/v1/zones/fr-par-2/ips/77e0cacb-6b15-4c65-b60f-47011b210fce
     method: GET
   response:
-    body: '{"ip": {"id": "07eae38f-a230-4586-9d58-1d1476f2d07b", "address": "51.159.151.134",
-      "reverse": null, "server": null, "organization": "ee7bd9e1-9cbd-4724-b2f4-19e50f3cf38b",
-      "project": "ee7bd9e1-9cbd-4724-b2f4-19e50f3cf38b", "zone": "fr-par-2", "tags":
-      []}}'
+    body: '{"ip":{"address":"51.159.152.92","id":"77e0cacb-6b15-4c65-b60f-47011b210fce","organization":"105bdce1-64c0-48ab-899d-868455867ecf","project":"105bdce1-64c0-48ab-899d-868455867ecf","reverse":null,"server":null,"tags":[],"zone":"fr-par-2"}}'
```
Comment (Monitob) [inline scaleway/testdata/provider-instance-ip-zones.cassette.yaml:220]: Couldn't find where this query is coming?

File: scaleway/resource_object_bucket_acl.go [hunk 12]
```diff
@@ -129,7 +129,8 @@ func resourceScalewayObjectBucketACL() *schema.Resource {
 				Description:  "The project ID as owner.",
 				ValidateFunc: validationUUID(),
 			},
-			"region": regionSchema(),
+			"region":     regionSchema(),
+			"project_id": projectIDSchema(),
```
Comment (Monitob) [inline scaleway/resource_object_bucket_acl.go:133]: Set is missing

File: scaleway/resource_object.go [hunk 13]
```diff
@@ -81,7 +81,8 @@ func resourceScalewayObject() *schema.Resource {
 					s3.ObjectCannedACLPublicRead,
 				}, false),
 			},
-			"region": regionSchema(),
+			"region":     regionSchema(),
+			"project_id": projectIDSchema(),
```
Comment (Monitob) [inline scaleway/resource_object.go:85]: Set is missing

File: scaleway/resource_object_bucket_lock_configuration.go [hunk 14]
```diff
@@ -69,6 +69,8 @@ func resourceObjectLockConfiguration() *schema.Resource {
 				},
 				Description: "Specifies the Object Lock rule for the specified object.",
 			},
+			"region":     regionSchema(),
+			"project_id": projectIDSchema(),
```
Comment (Monitob) [inline scaleway/resource_object_bucket_lock_configuration.go:73]: Set are missing

File: scaleway/resource_object_bucket_policy.go [hunk 15]
```diff
@@ -40,7 +40,8 @@ func resourceScalewayObjectBucketPolicy() *schema.Resource {
 				Description:      "The text of the policy.",
 				DiffSuppressFunc: SuppressEquivalentPolicyDiffs,
 			},
-			"region": regionSchema(),
+			"region":     regionSchema(),
+			"project_id": projectIDSchema(),
```
Comment (Monitob) [inline scaleway/resource_object_bucket_policy.go:44]: Set Missing

File: scaleway/resource_object_bucket_website_configuration.go [hunk 16]
```diff
@@ -71,6 +71,8 @@ func ResourceBucketWebsiteConfiguration() *schema.Resource {
 				Computed:    true,
 				Description: "The website endpoint.",
 			},
+			"region":     regionSchema(),
+			"project_id": projectIDSchema(),
```
Comment (Monitob) [inline scaleway/resource_object_bucket_website_configuration.go:75]: Set missing

Comment (Monitob) [inline scaleway/resource_object_bucket_website_configuration.go:75]: Maybe and example for this case. could be interesting.

**Review by Monitob — CHANGES_REQUESTED** [review]

File: scaleway/data_source_object_bucket_test.go (hunk 7 above)
Comment (Codelax) [reply scaleway/data_source_object_bucket_test.go:72]: This is tricks using a provider to do the test, we are implementing `project_id` to avoid doing this

File: scaleway/testdata/provider-instance-ip-zones.cassette.yaml (hunk 11 above)
Comment (quantumsheep) [reply scaleway/testdata/provider-instance-ip-zones.cassette.yaml:220]: It's from this: https://github.com/scaleway/terraform-provider-scaleway/blob/f5685ad6f550c0c815821408787dc0ae87f0f232/scaleway/provider_test.go#L362

**Review by Monitob — DISMISSED** [review]

**Review by Monitob — APPROVED** [review]

**Review by yfodil — APPROVED** [review]
